src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +362,80 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    protected async ValueTask TrackPrimordialOnlyProjectAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            // If project is already tracked, no need to do any further work.
```

Comment:
So the primordial cases are interesting -- if we ever had a situation where we already had a primordial project -- aren't we leaking one here? Because the caller will return thinking a later call to UnloadProjectAsync() can remove this project...but that won't be the case because we've "forgotten" about the ID.

Reply:
Thinking on it again I think we should throw if the project is already tracked. In some cases we could maybe keep things in a consistent state, but in general, if just some random `ProjectLoadState.LoadedTargets` is found in the dictionary, we don't have a way to make the states of the workspace and project system consistent at that point, things have gone badly wrong in that case.

Reply:
I think the approach you took is good: if we're just doing a regular load and it's already got a regular Loaded state, then that's sane since it's just queueing another refresh. But for primordial's it's a bit less certain, so blocking is fine.

(A conversation with @JoeRobich made me think we might be find some benefit of using the primordial stuff for the regular project system stuff too, maybe in that case we might structure this where the method here takes a lambda to create the primordial project if it's not already created. But we can cross that bridge when we get there.)